### PR TITLE
Add CreateFromMetadata in Paravirt PETM

### PR DESCRIPTION
**Pull request by xing-yang**
This adds support for restore in the guest cluster.

### CODE REVIEW

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 1]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
 
 	return data, md, combined, nil
 }
+
+// CreateFromMetadata creates CloneFromSnapshot CR in the Supervisor Cluster
+func (this *ParaVirtProtectedEntityTypeManager) CreateFromMetadata(ctx context.Context, metadata []byte, sourceSnapshotID astrolabe.ProtectedEntityID, componentSourcePETM astrolabe.ProtectedEntityTypeManager, cloneFromSnapshotNamespace string, cloneFromSnapshotName string, backupRepositoryName string) (astrolabe.ProtectedEntity, error) {
+	this.logger.Infof("CreateFromMeta called on Paravirtualized PETM. sourceSnapshotID: %s. cloneFromSnapshot: %s/%s, backupRepositoryName: %s", sourceSnapshotID.String(), cloneFromSnapshotNamespace, cloneFromSnapshotName, backupRepositoryName)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:192]: CreateFromMeta --> CreateFromMetadata

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 2]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+
+	// Get backupRepository from Guest, set backRepositoryName to backupRepositoryObj.SvcBackupRepositoryName
+	var backupRepo *snapshotUtils.BackupRepository
+	if backupRepositoryName != "" {
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:196]: do we need to check `backupRepositoryName != utils.WithoutBackupRepository` here?

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 3]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+		backupRepositoryCR, err := utils.GetBackupRepositoryFromBackupRepositoryName(backupRepositoryName)
+		if err != nil {
+			this.logger.Errorf("Failed to get BackupRepository from BackupRepositoryName %s: %v", backupRepositoryName, err)
+			return nil, errors.Wrapf(err, "failed to retrieve backupRepository")
+		}
+		if backupRepositoryCR.SvcBackupRepositoryName != "" {
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:202]: same as above.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 4]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	}
+
+	if svcClone.Status.Phase == backupdriverv1.ClonePhaseFailed {
+		this.logger.Errorf("CloneFromSnapshot CR %s/%s failed in the Supervisor Cluster", this.svcNamespace, svcCloneFromSnapshot.Name)
+		return nil, fmt.Errorf("CloneFromSnapshot failed in the Supervisor Cluster: %s/%s", svcClone.Namespace, svcClone.Name)
+	} else if svcClone.Status.Phase == backupdriverv1.ClonePhaseCanceled {
+		this.logger.Errorf("CloneFromSnapshot CR %s/%s is canceled in the Supervisor Cluster", this.svcNamespace, svcCloneFromSnapshot.Name)
+		return nil, fmt.Errorf("CloneFromSnapshot is canceled: %s/%s in the Supervisor Cluster", svcClone.Namespace, svcClone.Name)
+	}
+	this.logger.Infof("CreateFromMeta: CloneFromSnapshot %s/%s is completed in Supervisor Cluster. Phase: %v", svcClone.Namespace, svcClone.Name, svcClone.Status.Phase)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:276]: CreateFromMeta --> CreateFromMetadata

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 5]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	if pv, err = this.gcKubeClientSet.CoreV1().PersistentVolumes().Create(context.TODO(), pv, metav1.CreateOptions{}); err == nil || apierrs.IsAlreadyExists(err) {
+		// Save succeeded.
+		if err != nil {
+			this.logger.Infof("PV %s already exists, reusing", pv.Name)
+			err = nil
+		} else {
+			this.logger.Infof("PV %s saved", pv.Name)
+		}
+	}
+	this.logger.Infof("CreateFromMeta: PV %s created in the guest cluster", pv.Name)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:319]: CreateFromMeta --> CreateFromMetadata

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 6]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	if pvc, err = this.gcKubeClientSet.CoreV1().PersistentVolumeClaims(namespace).Create(context.TODO(), pvc, metav1.CreateOptions{}); err == nil || apierrs.IsAlreadyExists(err) {
+		// Save succeeded.
+		if err != nil {
+			this.logger.Infof("PVC %s/%s already exists, reusing", pvc.Namespace, pvc.Name)
+			err = nil
+		} else {
+			this.logger.Infof("PVC %s/%s saved", pvc.Namespace, pvc.Name)
+		}
+	}
+	this.logger.Infof("CreateFromMeta: PVC %s/%s is created in the guest cluster", pvc.Namespace, pvc.Name)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:348]: CreateFromMeta --> CreateFromMetadata

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 7]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	}
+
+	// Update CloneFromSnapshot status in Guest based on CloneFromSnapshot from Supervisor
+	clone := gcClone.DeepCopy()
+	clone.Status.Phase = svcClone.Status.Phase
+	clone.Status.Message = svcClone.Status.Message
+	clone.Status.ResourceHandle = svcClone.Status.ResourceHandle.DeepCopy()
+	_, err = this.gcBackupDriverClient.CloneFromSnapshots(cloneFromSnapshotNamespace).UpdateStatus(context.TODO(), clone, metav1.UpdateOptions{})
+	if err != nil {
+		this.logger.Errorf("CreateFromMeta: Failed to update status of CloneFromSnapshot %s/%s to %v", cloneFromSnapshotNamespace, cloneFromSnapshotName, clone.Status.Phase)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:372]: CreateFromMeta --> CreateFromMetadata

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 8]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+		return nil, err
+	}
+	this.logger.Infof("CreateFromMeta: CloneFromSnapshot %s/%s updated successfully to Phase %v", cloneFromSnapshotNamespace, cloneFromSnapshotName, clone.Status.Phase)
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:375]: CreateFromMeta --> CreateFromMetadata

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 9]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+
+	pe, err := this.GetProtectedEntity(ctx, peID)
+	if err != nil {
+		this.logger.WithError(err).Errorf("Failed to get the ProtectedEntity from peID %s", peID.String())
+		return nil, err
+	}
+
+	this.logger.Infof("CreateFromMetadata: retrieved ProtectedEntity for ID %s.", peID.String())
+	return pe, nil
+}
```
Comment (lintongj) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:388]: I would suggest to modularize this function implementation. It is a little bit long and contains a lot of detail. But, it doesn't have to happen in this PR.

Comment (xing-yang) [reply pkg/paravirt/paravirt_protected_entity_type_manager.go:388]: Good point.  Let me do it after this PR is merged.

**Comment by xing-yang** [comment]
Updated go.mod to pick up the latest astrolabe changes.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go (hunk 2 above)
Comment (xing-yang) [reply pkg/paravirt/paravirt_protected_entity_type_manager.go:196]: Sure.  Will add that.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 10]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+		}
+		if backupRepositoryCR.SvcBackupRepositoryName != "" {
+			this.logger.Info("BackupRepositoryName in Supervisor: %s", backupRepositoryCR.SvcBackupRepositoryName)
+			backupRepositoryName = backupRepositoryCR.SvcBackupRepositoryName
+		}
+		backupRepo = snapshotUtils.NewBackupRepository(backupRepositoryName)
+	}
+
+	// sourceSnapshotID from CloneFromSnapshot in Guest is in this format:
+	// pvc:test-ns-xtayual/test-pvc:cGFyYXZpcnQtcHY6cHZjLTE1ZDBhYzhmLTQxOGYtNGU1ZC05OWMxLWIxMjcyNjNlMjA1ODphWFprT2pGbE5EWmlZalJrTFdJelpqQXROREJrTlMwNVkyRTRMVE5pWVdVMlpqVTVOVGsxTlRwa01tVXlNamN4TmkwNVlUWXhMVFEyT0dFdFlqQmhZeTB4T1RKa01tSmtNV0kzWmpV
```
Comment (deepakkinni) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:210]: Remove these.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 11]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	if err != nil {
+		this.logger.WithError(err).Errorf("Failed to get the ProtectedEntity from peID %s", peID.String())
+		return nil, err
+	}
+
+	this.logger.Infof("CreateFromMetadata: retrieved ProtectedEntity for ID %s.", peID.String())
+	return pe, nil
+}
+
+func decodeSnapshotID(snapshotID astrolabe.ProtectedEntitySnapshotID, logger logrus.FieldLogger) (string, error) {
```
Comment (deepakkinni) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:390]: There's a decodeSnapshotID in snapMgr, please use that, i don't think there's much of difference. Better still, move it to utils.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 12]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	// and encode again before calling CreateFromMetadata
+	// NOTE(xyang): We need to make assumption that StorageClass is already created
+	// in Supervisor and Synchronized to Guest Cluster
+	// StorageClass in Guest must have StorageClass name of Supervisor
+	// parameters:
+	//   svStorageClass: gc-storage-profile
+	svcPVC := v1.PersistentVolumeClaim{}
+	err = svcPVC.Unmarshal(metadata)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to retrieve PVC")
```
Comment (deepakkinni) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:232]: This is not a failure in retrieving the PVC, failed to unmarshall?

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 13]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	this.logger.Infof("StorageClassName is %s in Supervisor PVC: %s/%s", *svcPVC.Spec.StorageClassName, svcPVC.Name, svcPVC.Namespace)
+	svcPVCData, err := svcPVC.Marshal()
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not marshal SVC PVC data for %v",
+			svcPVC.Name, svcPVC.Namespace)
+	}
+
+	// We need to construct the snapshotID in CloneFromSnapshot for Supervisor in this format:
+	// pvc:test-gc-e2e-demo-ns/test-pvc:aXZkOjFlNDZiYjRkLWIzZjAtNDBkNS05Y2E4LTNiYWU2ZjU5NTk1NTpkMmUyMjcxNi05YTYxLTQ2OGEtYjBhYy0xOTJkMmJkMWI3ZjU
+	// aXZkOjFlNDZiYjRkLWIzZjAtNDBkNS05Y2E4LTNiYWU2ZjU5NTk1NTpkMmUyMjcxNi05YTYxLTQ2OGEtYjBhYy0xOTJkMmJkMWI3ZjU will be decoded to ivd:1e46bb4d-b3f0-40d5-9ca8-3bae6f595955:ea4e347a-be29-4e5b-a626-725b83f168fcbase64 by PVC PETM
```
Comment (deepakkinni) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:244]: Remove.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 14]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	newSnapshotID := "pvc:" + svcPVC.Namespace + "/" + svcPVC.Name + ":" + snapIDDecoded
```
Comment (deepakkinni) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:245]: Create a util method to construct newSnapshotID in the desired format.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 15]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	svcCloneFromSnapshot, err := snapshotUtils.CloneFromSnapshopRef(ctx, this.svcBackupDriverClient, newSnapshotID, svcPVCData, &apiGroup, kind, this.svcNamespace, *backupRepo, []backupdriverv1.ClonePhase{backupdriverv1.ClonePhaseCompleted, backupdriverv1.ClonePhaseFailed, backupdriverv1.ClonePhaseCanceled}, this.logger)
+	this.logger.Infof("CreateFromMetadata: finished waiting for CloneFromSnapshot's status to be completed, failed, or canceled in the Supervisor Cluster")
+
+	if err != nil {
+		this.logger.Errorf("Failed to create a cloneFromSnapshot CR: %v", err)
+		return nil, err
+	}
+
+	// Get CloneFromSnapshot from Supervisor
+	svcClone, err := this.svcBackupDriverClient.CloneFromSnapshots(this.svcNamespace).Get(context.TODO(), svcCloneFromSnapshot.Name, metav1.GetOptions{})
```
Comment (deepakkinni) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:263]: Is there a value in retrieving this again? Considering the following lines just compare the phase, you could just use svcCloneFromSnapshot directly. Furthermore `snapshotUtils.CloneFromSnapshopRef` is a blocking call and won't return until it hits one of the status.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 16]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+	if err != nil {
+		this.logger.Errorf("Failed to get the cloneFromSnapshot CR %s/%s: %v", this.svcNamespace, svcCloneFromSnapshot.Name, err)
```
Comment (deepakkinni) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:265]: This is repeated several times, construct the string and use that string variable.

File: pkg/paravirt/paravirt_protected_entity_type_manager.go [hunk 17]
```diff
@@ -166,3 +186,230 @@ func (this *ParaVirtProtectedEntityTypeManager) getDataTransports(id astrolabe.P
...
+					Driver:       CSIDriverName,
+					VolumeHandle: svcPVC.Name,
+				},
+			},
+			ClaimRef: &v1.ObjectReference{
+				Namespace: namespace,
+				Name:      pvcName,
+			},
+		},
+	}
```
Comment (deepakkinni) [inline pkg/paravirt/paravirt_protected_entity_type_manager.go:308]: Move this out into it's own method.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/185/

**Comment by svcbot-qecnsdp** [comment]
Congratulations !! 

All the pre-check tests passed
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/186/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/187/

**Pull request merged**